hadoop-ozone/ozonefs/src/main/java/org/apache/hadoop/fs/ozone/OFSPath.java

```diff
@@ -238,4 +248,23 @@ static String getTempMountBucketNameOfCurrentUser() throws IOException {
     String username = UserGroupInformation.getCurrentUser().getUserName();
     return getTempMountBucketName(username);
   }
+
+  /**
+   * Return trash root for the given path.
+   * @return trash root for the given path
+   */
+  public Path getTrashRoot() {
+    try {
+      String username = UserGroupInformation.getCurrentUser().getUserName();
+      URI uri = new URIBuilder().setScheme(OZONE_OFS_URI_SCHEME)
```

Comment:
Should it be first username and then .Trash ? like ofs:///vol1/bucket1/username/.Trash  ?

This is default iml at FileSystem.java
```

 /**
   * Get the root directory of Trash for current user when the path specified
   * is deleted.
   *
   * @param path the trash root of the path to be determined.
   * @return the default implementation returns {@code /user/$USER/.Trash}
   */
  public Path getTrashRoot(Path path) {
    return this.makeQualified(new Path(getHomeDirectory().toUri().getPath(),
        TRASH_PREFIX));
  }
```
CC: @mukul1987 @sadanand48 @arp7 : I think this should be checked with Sadanand implementation as well to be in sync?

Reply:
Given we don't have a home directory per bucket, I think this PR took an approach similar to encryption zone where the .Trash is created per bucket, i.e., ofs://vol1/bucket1/.Trash/username, which have a common trash root per bucket. 

If we take ofs:///vol1/bucket1/username/.Trash, we won't have a common trash root. As a result, the trash clean up will be expensive as we don't have a common root per bucket to check the expired trash checkpoints.

Reply:
Thanks @xiaoyuyao for explanation.
My point is: after looking at Trash emptier impl demo today, in o3fs we seems to be getting <user>/.Trash
My point is, this o3fs and ofs trash directories should be in sync. so that emptier will not omit the  trash dirs created by o3fs?
Also inside OM Server, we get trashRoots from the fs configured as OM. That should default fs. When we configure ofs, it should also consider trash dirs created by o3fs? I think yes, so that they both should be in sync.

Reply:
I agree o3fs and ofs trash locations should be the same. We should refactor the common parts to some new shared classes later.

I didn't get the "OM Server" part though. IMO `getTrashRoot` is purely a client-side method?

Reply:
Yes its client side logic. TrashEmptier uses fs to get the directories and this TrashEmptier runs inside NN, so here it will be in OM server.
If we wanted to reuse the logic, all dirs should be in sync and TrashEmptier will get the trashRoots and cleanup.

For your ref below is the TrashEmptier code: 
 ```
 try {
          now = Time.now();
          if (now >= end) {
            Collection<FileStatus> trashRoots;
            trashRoots = fs.getTrashRoots(true);      // list all trash dirs

            for (FileStatus trashRoot : trashRoots) {   // dump each trash
              if (!trashRoot.isDirectory())
                continue;
              try {
                TrashPolicyDefault trash = new TrashPolicyDefault(fs, conf);
                trash.deleteCheckpoint(trashRoot.getPath(), false);
                trash.createCheckpoint(trashRoot.getPath(), new Date(now));
              } catch (IOException e) {
                LOG.warn("Trash caught: "+e+". Skipping " +
                    trashRoot.getPath() + ".");
              } 
            }
          }
        } catch (Exception e) {
          LOG.warn("RuntimeException during Trash.Emptier.run(): ", e); 
        }
```

Reply:
I am in favor of maintaining trash roots for ofs and o3fs the same, as @umamaheswararao said in the first phase, we can reuse the code of hdfs logic and run the same in OM with OFS filesystem object and also for this we need to implement fs.getTrashRoots return path for all buckets in the system(Which can be done in a separate Jira).

Reply:
@umamaheswararao I think we are on the same page. Both o3fs and ofs have Trash per bucket. And code should be reused.